---
title: Arrêt des erreurs | Microsoft Docs
ms.custom: ''
ms.date: 09/13/2016
ms.reviewer: ''
ms.suite: ''
ms.tgt_pltfrm: ''
ms.topic: article
ms.assetid: b804e738-aefa-41bb-9649-f9ed897fd96c
caps.latest.revision: 8
ms.openlocfilehash: d1967fe7996f75ec5229920f7ec49aa5ff6bdbfd
ms.sourcegitcommit: debd2b38fb8070a7357bf1a4bf9cc736f3702f31
ms.translationtype: MT
ms.contentlocale: fr-FR
ms.lasthandoff: 12/05/2019
ms.locfileid: "72369328"
---
# <a name="terminating-errors"></a><span data-ttu-id="1cfde-102">Erreur avec fin d’exécution</span><span class="sxs-lookup"><span data-stu-id="1cfde-102">Terminating Errors</span></span>

<span data-ttu-id="1cfde-103">Cette rubrique décrit la méthode utilisée pour signaler des erreurs de fin.</span><span class="sxs-lookup"><span data-stu-id="1cfde-103">This topic discusses the method used to report terminating errors.</span></span> <span data-ttu-id="1cfde-104">Elle explique également comment appeler la méthode à partir de l’applet de commande et traite des exceptions qui peuvent être retournées par le runtime Windows PowerShell lorsque la méthode est appelée.</span><span class="sxs-lookup"><span data-stu-id="1cfde-104">It also discusses how to call the method from within the cmdlet, and it discusses the exceptions that can be returned by the Windows PowerShell runtime when the method is called.</span></span>

<span data-ttu-id="1cfde-105">Lorsqu’une erreur se termine, l’applet de commande doit signaler l’erreur en appelant la méthode [System. Management. Automation. applet de commande. ThrowTerminatingError \*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) .</span><span class="sxs-lookup"><span data-stu-id="1cfde-105">When a terminating error occurs, the cmdlet should report the error by calling the [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) method.</span></span> <span data-ttu-id="1cfde-106">Cette méthode permet à l’applet de commande d’envoyer un enregistrement d’erreur qui décrit la condition à l’origine de l’erreur.</span><span class="sxs-lookup"><span data-stu-id="1cfde-106">This method allows the cmdlet to send an error record that describes the condition that caused the terminating error.</span></span> <span data-ttu-id="1cfde-107">Pour plus d’informations sur les enregistrements d’erreurs, consultez la page [enregistrements d’erreurs Windows PowerShell](./windows-powershell-error-records.md).</span><span class="sxs-lookup"><span data-stu-id="1cfde-107">For more information about error records, see [Windows PowerShell Error Records](./windows-powershell-error-records.md).</span></span>

<span data-ttu-id="1cfde-108">Lorsque la méthode [System. Management. Automation. applet de commande. ThrowTerminatingError \*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) est appelée, le runtime Windows PowerShell arrête définitivement l’exécution du pipeline et lève une exception [System. Management. Automation. Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) .</span><span class="sxs-lookup"><span data-stu-id="1cfde-108">When the [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) method is called, the  Windows PowerShell runtime permanently stops the execution of the pipeline and throws a [System.Management.Automation.Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) exception.</span></span> <span data-ttu-id="1cfde-109">Toutes les tentatives ultérieures pour appeler [System. Management. Automation. applet de commande. WriteObject](/dotnet/api/System.Management.Automation.Cmdlet.WriteObject), [System. Management. Automation. applet](/dotnet/api/System.Management.Automation.Cmdlet.WriteError)de commande. WriteError ou plusieurs autres API provoquent la levée d’une exception [System. Management. Automation. Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) par les appels.</span><span class="sxs-lookup"><span data-stu-id="1cfde-109">Any subsequent attempts to call [System.Management.Automation.Cmdlet.WriteObject](/dotnet/api/System.Management.Automation.Cmdlet.WriteObject), [System.Management.Automation.Cmdlet.WriteError](/dotnet/api/System.Management.Automation.Cmdlet.WriteError), or several other APIs causes those calls to throw a [System.Management.Automation.Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) exception.</span></span>

<span data-ttu-id="1cfde-110">L’exception [System. Management. Automation. Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) peut également se produire si une autre applet de commande du pipeline signale une erreur de fin, si l’utilisateur a demandé l’arrêt du pipeline, ou si le pipeline a été interrompu avant la fin pour une raison quelconque.</span><span class="sxs-lookup"><span data-stu-id="1cfde-110">The [System.Management.Automation.Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) exception can also occur if another cmdlet in the pipeline reports a terminating error, if the user has asked to stop the pipeline, or if the pipeline has been halted before completion for any reason.</span></span> <span data-ttu-id="1cfde-111">L’applet de commande n’a pas besoin d’intercepter l’exception [System. Management. Automation. Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) , sauf si elle doit nettoyer les ressources ouvertes ou son état interne.</span><span class="sxs-lookup"><span data-stu-id="1cfde-111">The cmdlet does not need to catch the [System.Management.Automation.Pipelinestoppedexception](/dotnet/api/System.Management.Automation.PipelineStoppedException) exception unless it must clean up open resources or its internal state.</span></span>

<span data-ttu-id="1cfde-112">Les applets de commande peuvent écrire un nombre quelconque d’objets de sortie ou d’erreurs sans fin d’achèvement avant de signaler une erreur de fin.</span><span class="sxs-lookup"><span data-stu-id="1cfde-112">Cmdlets can write any number of output objects or non-terminating errors before reporting a terminating error.</span></span> <span data-ttu-id="1cfde-113">Toutefois, l’erreur d’arrêt interrompt définitivement le pipeline, et aucune sortie supplémentaire, les erreurs de fin ou les erreurs sans fin d’arrêt peuvent être signalées.</span><span class="sxs-lookup"><span data-stu-id="1cfde-113">However, the terminating error permanently stops the pipeline, and no further output, terminating errors, or non-terminating errors can be reported.</span></span>

<span data-ttu-id="1cfde-114">Les applets de commande peuvent appeler [System. Management. Automation. applet de commande. ThrowTerminatingError \*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) uniquement à partir du thread qui a appelé la méthode de traitement d’entrée System. Management. Automation. cmdlet. [BeginProcessing](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing), [System. Management. Automation. cmdlet. ProcessRecord](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord)ou [System. Management. Automation. applet de commande. EndProcessing](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing) .</span><span class="sxs-lookup"><span data-stu-id="1cfde-114">Cmdlets can call [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) only from the thread that called the [System.Management.Automation.Cmdlet.BeginProcessing](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing), [System.Management.Automation.Cmdlet.ProcessRecord](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord), or [System.Management.Automation.Cmdlet.EndProcessing](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing) input processing method.</span></span> <span data-ttu-id="1cfde-115">N’essayez pas d’appeler [System. Management. Automation. applet](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) de commande. ThrowTerminatingError \* ou [System. Management. Automation. cmdlet. WriteError](/dotnet/api/System.Management.Automation.Cmdlet.WriteError) à partir d’un autre thread.</span><span class="sxs-lookup"><span data-stu-id="1cfde-115">Do not attempt to call [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) or [System.Management.Automation.Cmdlet.WriteError](/dotnet/api/System.Management.Automation.Cmdlet.WriteError) from another thread.</span></span> <span data-ttu-id="1cfde-116">Au lieu de cela, les erreurs doivent être communiquées au thread principal.</span><span class="sxs-lookup"><span data-stu-id="1cfde-116">Instead, errors must be communicated back to the main thread.</span></span>

<span data-ttu-id="1cfde-117">Une applet de commande peut lever une exception dans son implémentation de la méthode System. [Management. Automation. applet de commande. BeginProcessing](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing), [System. Management. Automation. cmdlet. ProcessRecord](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord)ou [System. Management. Automation. cmdlet. EndProcessing](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing) .</span><span class="sxs-lookup"><span data-stu-id="1cfde-117">It is possible for a cmdlet to throw an exception in its implementation of the [System.Management.Automation.Cmdlet.BeginProcessing](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing), [System.Management.Automation.Cmdlet.ProcessRecord](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord), or [System.Management.Automation.Cmdlet.EndProcessing](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing) method.</span></span> <span data-ttu-id="1cfde-118">Toute exception levée à partir de ces méthodes (à l’exception de quelques conditions d’erreur graves qui arrêtent l’hôte Windows PowerShell) est interprétée comme une erreur de fin qui arrête le pipeline, mais pas Windows PowerShell dans son ensemble.</span><span class="sxs-lookup"><span data-stu-id="1cfde-118">Any exception thrown from these methods (except for a few severe error conditions that stop the Windows PowerShell host) is interpreted as a terminating error which stops the pipeline, but not Windows PowerShell as a whole.</span></span> <span data-ttu-id="1cfde-119">(Cela s’applique uniquement au thread principal de l’applet de commande.</span><span class="sxs-lookup"><span data-stu-id="1cfde-119">(This applies only to the main cmdlet thread.</span></span> <span data-ttu-id="1cfde-120">Exceptions non interceptées dans les threads générés par l’applet de commande, en général, arrêtez l’hôte Windows PowerShell.) Nous vous recommandons d’utiliser [System. Management. Automation. applet de commande. ThrowTerminatingError \*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) au lieu de lever une exception, car l’enregistrement d’erreur fournit des informations supplémentaires sur la condition d’erreur, ce qui est utile pour l’utilisateur final.</span><span class="sxs-lookup"><span data-stu-id="1cfde-120">Uncaught exceptions in threads spawned by the cmdlet, in general, halt the Windows PowerShell host.) We recommend that you use [System.Management.Automation.Cmdlet.Throwterminatingerror\*](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError) rather than throwing an exception because the error record provides additional information about the error condition, which is useful to the end-user.</span></span> <span data-ttu-id="1cfde-121">Les applets de commande doivent respecter les instructions du code managé par rapport à l’interception et à la gestion de toutes les exceptions (`catch (Exception e)`).</span><span class="sxs-lookup"><span data-stu-id="1cfde-121">Cmdlets should honor the managed code guideline against catching and handling all exceptions (`catch (Exception e)`).</span></span> <span data-ttu-id="1cfde-122">Convertit uniquement les exceptions des types connus et attendus en enregistrements d’erreur.</span><span class="sxs-lookup"><span data-stu-id="1cfde-122">Convert only exceptions of known and expected types into error records.</span></span>

## <a name="see-also"></a><span data-ttu-id="1cfde-123">Voir aussi</span><span class="sxs-lookup"><span data-stu-id="1cfde-123">See Also</span></span>

[<span data-ttu-id="1cfde-124">System. Management. Automation. applet de commande. BeginProcessing</span><span class="sxs-lookup"><span data-stu-id="1cfde-124">System.Management.Automation.Cmdlet.BeginProcessing</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.BeginProcessing)

[<span data-ttu-id="1cfde-125">System. Management. Automation. applet de commande. EndProcessing</span><span class="sxs-lookup"><span data-stu-id="1cfde-125">System.Management.Automation.Cmdlet.EndProcessing</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.EndProcessing)

[<span data-ttu-id="1cfde-126">System. Management. Automation. applet de commande. ProcessRecord</span><span class="sxs-lookup"><span data-stu-id="1cfde-126">System.Management.Automation.Cmdlet.ProcessRecord</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.ProcessRecord)

[<span data-ttu-id="1cfde-127">System. Management. Automation. Pipelinestoppedexception</span><span class="sxs-lookup"><span data-stu-id="1cfde-127">System.Management.Automation.Pipelinestoppedexception</span></span>](/dotnet/api/System.Management.Automation.PipelineStoppedException)

[<span data-ttu-id="1cfde-128">System. Management. Automation. applet de commande. ThrowTerminatingError \*</span><span class="sxs-lookup"><span data-stu-id="1cfde-128">System.Management.Automation.Cmdlet.Throwterminatingerror\*</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.ThrowTerminatingError)

[<span data-ttu-id="1cfde-129">System. Management. Automation. applet de commande. WriteError</span><span class="sxs-lookup"><span data-stu-id="1cfde-129">System.Management.Automation.Cmdlet.WriteError</span></span>](/dotnet/api/System.Management.Automation.Cmdlet.WriteError)

[<span data-ttu-id="1cfde-130">Enregistrements d’erreurs Windows PowerShell</span><span class="sxs-lookup"><span data-stu-id="1cfde-130">Windows PowerShell Error Records</span></span>](./windows-powershell-error-records.md)

[<span data-ttu-id="1cfde-131">Écriture d’une applet de commande Windows PowerShell</span><span class="sxs-lookup"><span data-stu-id="1cfde-131">Writing a Windows PowerShell Cmdlet</span></span>](./writing-a-windows-powershell-cmdlet.md)
